File: src/config/env.ts
import "dotenv/config";

import path from "node:path";

import { z } from "zod";

import {
  ConfigError,
  defaultSettingsPath,
  expandHomePath,
  loadGoogleFitCredentials,
  loadOrCreateSettings,
  type DefaultSource,
  type SettingsFile,
} from "./settings.js";

export { ConfigError, expandHomePath } from "./settings.js";
export type { DefaultSource } from "./settings.js";

export const defaultGoogleSheetId =
  "163Lc0YeN5ZnGeXdYqx6T_JGSMa91kpvfpoODjF7q8C0";
export const defaultGoogleSheetName = "体温・血圧";

const optionalString = z
  .string()
  .trim()
  .transform((value) => (value.length === 0 ? undefined : value))
  .optional();

const defaultedString = (defaultValue: string) =>
  z.preprocess(
    (value) =>
      typeof value === "string" && value.trim().length === 0
        ? undefined
        : value,
    z.string().trim().min(1).default(defaultValue),
  );

const positiveIntegerFromString = z
  .string()
  .trim()
  .regex(/^\d+$/)
  .transform((value) => Number.parseInt(value, 10))
  .pipe(z.number().int().positive());

export const envSchema = z.object({
  TIME_ZONE: z.string().trim().min(1).default("Asia/Tokyo"),
  GOOGLE_SHEET_ID: defaultedString(defaultGoogleSheetId),
  GOOGLE_SHEET_NAME: defaultedString(defaultGoogleSheetName),
  GOOGLE_APPLICATION_CREDENTIALS: optionalString,
  GOOGLE_FIT_CLIENT_ID: optionalString,
  GOOGLE_FIT_CLIENT_SECRET: optionalString,
  GOOGLE_FIT_REDIRECT_URI: z
    .string()
    .trim()
    .url()
    .default("http://localhost:3000/oauth2callback"),
  GOOGLE_FIT_TOKEN_PATH: z
    .string()
    .trim()
    .min(1)
    .default("~/.config/scale2sheet/google-fit-token.json"),
  GOOGLE_FIT_LOOKBACK_DAYS: positiveIntegerFromString.default(14),
  APPLE_HEALTH_EXPORT_XML: optionalString,
  SCALE_EXPORTER_OUTPUT_DIR: defaultedString("~/Documents/scale_exporter"),
  MORNING_CRON: z.string().trim().min(1).default("0 7 * * *"),
  EVENING_CRON: z.string().trim().min(1).default("0 21 * * *"),
});

export type EnvConfig = z.infer<typeof envSchema>;

export interface GoogleFitAuthConfig {
  readonly clientId: string;
  readonly clientSecret: string;
  readonly redirectUri: string;
  readonly tokenPath: string;
  readonly lookbackDays: number;
}

export interface GoogleSheetsAuthConfig {
  readonly applicationCredentialsPath: string;
  readonly spreadsheetId: string;
  readonly sheetName: string;
}

export interface AppleHealthConfig {
  readonly exportXmlPath: string;
}

export interface ScaleExporterConfig {
  readonly outputDir: string;
}

export interface SchedulerConfig {
  readonly morningCron: string;
  readonly eveningCron: string;
}

export interface AppConfig {
  readonly timeZone: string;
  readonly defaultSource: DefaultSource;
  readonly googleFit?: GoogleFitAuthConfig;
  readonly googleSheets?: GoogleSheetsAuthConfig;
  readonly appleHealth?: AppleHealthConfig;
  readonly scaleExporter: ScaleExporterConfig;
  readonly scheduler: SchedulerConfig;
}

type MutableAppConfig = {
  -readonly [Key in keyof AppConfig]: AppConfig[Key];
};

export interface LoadConfigOptions {
  /**
   * settings.json のパス。省略時は ~/.config/scale2sheet/settings.json
   * （なければ自動生成）。null で設定ファイル層を無効化（テスト用）。
   */
  readonly settingsPath?: string | null;
}

// settings.json のキー → 環境変数名の対応（環境変数が優先）
function settingsAsEnvOverlay(settings: SettingsFile): Record<string, string> {
  const overlay: Record<string, string> = {};
  const mapping: ReadonlyArray<[keyof SettingsFile, string]> = [
    ["time-zone", "TIME_ZONE"],
    ["sheet-id", "GOOGLE_SHEET_ID"],
    ["sheet-name", "GOOGLE_SHEET_NAME"],
    ["sheets-credentials", "GOOGLE_APPLICATION_CREDENTIALS"],
    ["scale-exporter-output-dir", "SCALE_EXPORTER_OUTPUT_DIR"],
    ["apple-health-export-xml", "APPLE_HEALTH_EXPORT_XML"],
    ["google-fit-token-path", "GOOGLE_FIT_TOKEN_PATH"],
    ["google-fit-lookback-days", "GOOGLE_FIT_LOOKBACK_DAYS"],
    ["morning-cron", "MORNING_CRON"],
    ["evening-cron", "EVENING_CRON"],
  ];

  for (const [settingsKey, envKey] of mapping) {
    const value = settings[settingsKey];
    if (value !== undefined) {
      overlay[envKey] = String(value);
    }
  }

  return overlay;
}

export function loadConfig(
  env: NodeJS.ProcessEnv = process.env,
  options: LoadConfigOptions = {},
): AppConfig {
  const settingsPath =
    options.settingsPath === undefined
      ? defaultSettingsPath
      : options.settingsPath;
  const settings: SettingsFile =
    settingsPath === null ? {} : loadOrCreateSettings(settingsPath);
  const configDir =
    settingsPath === null
      ? null
      : path.dirname(expandHomePath(settingsPath));

  // 優先順位: 環境変数（空文字は未設定扱い） > settings.json > 既定値
  const merged: Record<string, string> = settingsAsEnvOverlay(settings);
  for (const key of Object.keys(envSchema.shape)) {
    const value = env[key];
    if (typeof value === "string" && value.trim().length > 0) {
      merged[key] = value;
    }
  }

  const parsed = envSchema.parse(merged);
  const config: MutableAppConfig = {
    timeZone: parsed.TIME_ZONE,
    defaultSource: settings.source ?? "scale-exporter",
    scaleExporter: {
      outputDir: expandHomePath(parsed.SCALE_EXPORTER_OUTPUT_DIR),
    },
    scheduler: {
      morningCron: parsed.MORNING_CRON,
      eveningCron: parsed.EVENING_CRON,
    },
  };

  let clientId = parsed.GOOGLE_FIT_CLIENT_ID;
  let clientSecret = parsed.GOOGLE_FIT_CLIENT_SECRET;
  let redirectUri = parsed.GOOGLE_FIT_REDIRECT_URI;
  if ((!clientId || !clientSecret) && configDir !== null) {
    const credentials = loadGoogleFitCredentials(configDir);
    if (credentials) {
      clientId = credentials.clientId;
      clientSecret = credentials.clientSecret;
      redirectUri = credentials.redirectUri ?? redirectUri;
    }
  }
  if (clientId && clientSecret) {
    config.googleFit = {
      clientId,
      clientSecret,
      redirectUri,
      tokenPath: expandHomePath(parsed.GOOGLE_FIT_TOKEN_PATH),
      lookbackDays: parsed.GOOGLE_FIT_LOOKBACK_DAYS,
    };
  }

  if (parsed.GOOGLE_APPLICATION_CREDENTIALS && parsed.GOOGLE_SHEET_ID) {
    config.googleSheets = {
      applicationCredentialsPath: expandHomePath(
        parsed.GOOGLE_APPLICATION_CREDENTIALS,
      ),
      spreadsheetId: parsed.GOOGLE_SHEET_ID,
      sheetName: parsed.GOOGLE_SHEET_NAME,
    };
  }

  if (parsed.APPLE_HEALTH_EXPORT_XML) {
    config.appleHealth = {
      exportXmlPath: expandHomePath(parsed.APPLE_HEALTH_EXPORT_XML),
    };
  }

  return config;
}

export function requireGoogleFitConfig(
  config: AppConfig,
): GoogleFitAuthConfig {
  if (!config.googleFit) {
    throw new ConfigError(
      "Google Fit requires client credentials: set GOOGLE_FIT_CLIENT_ID / GOOGLE_FIT_CLIENT_SECRET, or create ~/.config/scale2sheet/google-fit-credentials.json.",
    );
  }

  return config.googleFit;
}

export function requireGoogleSheetsConfig(
  config: AppConfig,
): GoogleSheetsAuthConfig {
  if (!config.googleSheets) {
    throw new ConfigError(
      "Google Sheets requires credentials: set sheets-credentials in settings.json or GOOGLE_APPLICATION_CREDENTIALS.",
    );
  }

  return config.googleSheets;
}

export function requireAppleHealthConfig(config: AppConfig): AppleHealthConfig {
  if (!config.appleHealth) {
    throw new ConfigError(
      "Apple Health requires apple-health-export-xml in settings.json or APPLE_HEALTH_EXPORT_XML.",
    );
  }

  return config.appleHealth;
}

File: internal/config/env.go
package config

import (
	"fmt"
	"os"
	"path/filepath"
	"strconv"
	"strings"
)

const DefaultGoogleSheetName = "体温・血圧"

type GoogleFitAuthConfig struct {
	ClientID     string
	ClientSecret string
	RedirectURI  string
	TokenPath    string
	LookbackDays int
}

type GoogleSheetsAuthConfig struct {
	ApplicationCredentialsPath string
	SpreadsheetID              string
	SheetName                  string
}

type AppleHealthConfig struct {
	ExportXMLPath string
}

type ScaleExporterConfig struct {
	OutputDir string
}

type SchedulerConfig struct {
	MorningCron string
	EveningCron string
}

type AppConfig struct {
	TimeZone      string
	DefaultSource string
	GoogleFit     *GoogleFitAuthConfig
	GoogleSheets  *GoogleSheetsAuthConfig
	AppleHealth   *AppleHealthConfig
	ScaleExporter *ScaleExporterConfig
	Scheduler     SchedulerConfig
}

func Load(settingsPath string, env map[string]string) (AppConfig, error) {
	homeDir, err := os.UserHomeDir()
	if err != nil {
		return AppConfig{}, &ConfigError{Message: fmt.Sprintf("cannot resolve home directory: %v", err)}
	}
	if settingsPath == "" {
		settingsPath = DefaultSettingsPath
	}
	resolvedSettingsPath := ExpandHomePath(settingsPath, homeDir)
	settings, err := LoadOrCreateSettings(resolvedSettingsPath)
	if err != nil {
		return AppConfig{}, err
	}
	configDir := filepath.Dir(resolvedSettingsPath)

	getString := func(key string, setting *string, fallback string) string {
		if value := nonBlank(env[key]); value != "" {
			return value
		}
		if setting != nil {
			return strings.TrimSpace(*setting)
		}
		return fallback
	}

	config := AppConfig{
		TimeZone:      getString("TIME_ZONE", settings.TimeZone, "Asia/Tokyo"),
		DefaultSource: "scale-exporter",
		Scheduler: SchedulerConfig{
			MorningCron: getString("MORNING_CRON", settings.MorningCron, "0 7 * * *"),
			EveningCron: getString("EVENING_CRON", settings.EveningCron, "0 21 * * *"),
		},
	}
	if settings.Source != nil {
		config.DefaultSource = *settings.Source
	}

	if outputDir := getString("SCALE_EXPORTER_OUTPUT_DIR", settings.ScaleExporterOutputDir, ""); outputDir != "" {
		config.ScaleExporter = &ScaleExporterConfig{OutputDir: ExpandHomePath(outputDir, homeDir)}
	}
	if exportXML := getString("APPLE_HEALTH_EXPORT_XML", settings.AppleHealthExportXML, ""); exportXML != "" {
		config.AppleHealth = &AppleHealthConfig{ExportXMLPath: ExpandHomePath(exportXML, homeDir)}
	}

	sheetID := getString("GOOGLE_SHEET_ID", settings.SheetID, "")
	credentialsPath := getString("GOOGLE_APPLICATION_CREDENTIALS", settings.SheetsCredentials, "")
	if sheetID != "" && credentialsPath != "" {
		config.GoogleSheets = &GoogleSheetsAuthConfig{
			ApplicationCredentialsPath: ExpandHomePath(credentialsPath, homeDir),
			SpreadsheetID:              sheetID,
			SheetName:                  getString("GOOGLE_SHEET_NAME", settings.SheetName, DefaultGoogleSheetName),
		}
	}

	clientID := getString("GOOGLE_FIT_CLIENT_ID", settings.GoogleFitClientID, "")
	clientSecret := getString("GOOGLE_FIT_CLIENT_SECRET", settings.GoogleFitClientSecret, "")
	redirectURI := getString("GOOGLE_FIT_REDIRECT_URI", settings.GoogleFitRedirectURI, "http://localhost:3000/oauth2callback")
	if (clientID == "" || clientSecret == "") && configDir != "" {
		credentials, err := LoadGoogleFitCredentials(configDir)
		if err != nil {
			return AppConfig{}, err
		}
		if credentials != nil {
			clientID = credentials.ClientID
			clientSecret = credentials.ClientSecret
			if credentials.RedirectURI != "" {
				redirectURI = credentials.RedirectURI
			}
		}
	}
	lookbackDays, err := positiveIntSetting("GOOGLE_FIT_LOOKBACK_DAYS", settings.GoogleFitLookbackDays, env, 14)
	if err != nil {
		return AppConfig{}, err
	}
	if clientID != "" && clientSecret != "" {
		config.GoogleFit = &GoogleFitAuthConfig{
			ClientID:     clientID,
			ClientSecret: clientSecret,
			RedirectURI:  redirectURI,
			TokenPath:    ExpandHomePath(getString("GOOGLE_FIT_TOKEN_PATH", settings.GoogleFitTokenPath, "~/.config/scale2sheet/google-fit-token.json"), homeDir),
			LookbackDays: lookbackDays,
		}
	}

	return config, nil
}

func RequireGoogleFitConfig(config AppConfig) (*GoogleFitAuthConfig, error) {
	if config.GoogleFit == nil {
		return nil, &ConfigError{Message: "Google Fit requires client credentials: set GOOGLE_FIT_CLIENT_ID / GOOGLE_FIT_CLIENT_SECRET, or create ~/.config/scale2sheet/google-fit-credentials.json."}
	}
	return config.GoogleFit, nil
}

func RequireGoogleSheetsConfig(config AppConfig) (*GoogleSheetsAuthConfig, error) {
	if config.GoogleSheets == nil {
		return nil, &ConfigError{Message: "Google Sheets requires both sheet-id and sheets-credentials in ~/.config/scale2sheet/settings.json (or GOOGLE_SHEET_ID / GOOGLE_APPLICATION_CREDENTIALS)."}
	}
	return config.GoogleSheets, nil
}

func RequireAppleHealthConfig(config AppConfig) (*AppleHealthConfig, error) {
	if config.AppleHealth == nil {
		return nil, &ConfigError{Message: "Apple Health requires apple-health-export-xml in settings.json or APPLE_HEALTH_EXPORT_XML."}
	}
	return config.AppleHealth, nil
}

func RequireScaleExporterConfig(config AppConfig) (*ScaleExporterConfig, error) {
	if config.ScaleExporter == nil {
		return nil, &ConfigError{Message: "The scale-exporter source requires scale-exporter-output-dir in ~/.config/scale2sheet/settings.json (or SCALE_EXPORTER_OUTPUT_DIR), pointing at your scale_exporter JSONL output folder."}
	}
	return config.ScaleExporter, nil
}

func nonBlank(value string) string { return strings.TrimSpace(value) }

func positiveIntSetting(key string, setting *int, env map[string]string, fallback int) (int, error) {
	if value := nonBlank(env[key]); value != "" {
		parsed, err := strconv.Atoi(value)
		if err != nil || parsed <= 0 {
			return 0, &ConfigError{Message: fmt.Sprintf("invalid %s: must be a positive integer", key)}
		}
		return parsed, nil
	}
	if setting != nil {
		if *setting <= 0 {
			return 0, &ConfigError{Message: "invalid google-fit-lookback-days: must be positive"}
		}
		return *setting, nil
	}
	return fallback, nil
}
